Polkaholic.io's Multi-Chain Substrate Block Explorer

sourabhniyogi opened the pull request:
### Project Abstract

Polkaholic.io is a Substrate block explorer with high parachain coverage (45+ chains) and is developing multi-chain functionality for the next generation of Polkadot users.  The Polkaholic.io team would like to seek Web3F Level 2 funding for high availability XCM APIs powering the Polkaholic.io UI.  These XCM APIs will support Polkadot parachain and dapp developers, and will be used in the Polkaholic.io UI to support useful multi-chain block explorer features.

### For which grant [level](https://github.com/w3f/Grants-Program#level_slider-levels) are you applying?

- [ ] **Level 1**:  Up to $10,000, 2 approvals
- [x] **Level 2**:  Up to $30,000, 3 approvals
- [ ] **Level 3**:  Unlimited, 5 approvals (for > $100k Web3 Foundation Council approval)

### Application Checklist

- [x] The [application template](https://github.com/w3f/Grants-Program/blob/master/applications/application-template.md) has been copied, renamed ( `project_name.md`) and updated.
- [x] I have read and understood the [FAQs](https://github.com/w3f/Grants-Program/blob/master/docs/faq.md), [application guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/grant_guidelines_per_category.md) and [announcement guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/announcement-guidelines.md).
- [x] A BTC, Ethereum (USDT/USDC/DAI) or Polkadot/Kusama (aUSD) address for the payment of the milestones is provided inside the application.
- [x] I have read and acknowledge the [terms and conditions](https://github.com/w3f/Grants-Program/blob/master/docs/T&Cs.md).
- [x] The software delivered for this grant will be released under an open-source license specified in the application.
- [x] The initial PR contains only one commit (squash and force-push if needed).
- [x] The grant will only be announced once the first milestone [has been accepted](https://github.com/w3f/Grant-Milestone-Delivery#process).
- [x] I prefer the discussion of this application to be in a private Element/Matrix channel. My username is: ``@sourabhniyogi:matrix.org``

### How Did You Hear About our grants program?

- [ ] Social Media
- [ ] Hackathon
- [x] Personal Recommendation
- [ ] Substrate Builders Program
- [ ] Investor/VC
- [ ] Online Search
- [ ] Other: _______

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/w3f/Grants-Program?pullRequest=1039) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/w3f/Grants-Program?pullRequest=1039) before we can accept your contribution.<br/><hr/>**Sourabh Niyogi** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/w3f/Grants-Program?pullRequest=1039) it.</sub>

sourabhniyogi commented:
@Noc2 -- We are happy to open source 100% of code powering [polkaholic.io](http://polkaholic.io/) with a few caveats:
 (a) we have a Google Cloud centric stack (heavy use of BigTable) so anyone reusing the core will have to commit to that
 (b) actually getting the index requires fetching O(10^7) [82MM] traces/blocks/... from archive nodes supporting "unsafe" calls [for which we run 10-15 ourselves , the rest we get from onfinality in a semi-supervised way] -- we haven't started from 0 more than once
 (c) above indexing processes do not fit into typical single Docker container deployments in any meaningful way, which appears to be a requirement for all grants?   We can, however, have the UI in a single Docker container using the [polkaholic.io](http://polkaholic.io/) APIs.
We thought a self-contained Docker XCM-centric indexer + associated APIs (without any Google Cloud dependencies, just mysql) would be a way to have a "small code base" deliverable that (unlike all the indexing processes) could be easily understood / replicated and satisfy the Docker deliverable.
Can you advise on how to alter the scope given the above caveats?

Noc2 reviewed (changes requested):
Thanks for the quick reply. I would suggest changing the scope of the project to open source development that you plan to do or partly already did. As part of your first delivery you can open source the entire project. Maybe the second milestone could be UI focused or improving the existing backend etc. The main reason why we want teams to submit docker images is so that it’s easier for us to test it (especially with all the rust and substrate versions), but this isn’t a key requirement for grants and if it makes no sense in your case, feel free to not add it to the deliverables. Regarding indexing, we just need to find a way to test your deliverables, maybe we only index a few blocks or you share some data with us in private, etc. 
Alternatively you could also open source only parts that might be useful for others or smaller solutions, but I’m not sure it makes much sense for you to focus on “an additional” product. I would focus on improving your current solution. Btw. this way, you can receive grant funding for the development of the project and also apply for treasury support to actually run it (similar to polkascan).

sourabhniyogi commented:
We have released Polkaholic.io code under a GPLv3 license at
 https://github.com/colorfulnotion/polkaholic 
and revised to proposal to have the first 2 milestones focused on *verifiability* of results not with proofs but in the simpler sense of "does the indexed results match what you get from RPC calls"?   This will:
 (1) result in automated testing of the polkaholic.io UI/API 
 (2) increase user trust with browser-based JS + server-based JS (because finding bugs in indexer code is hard but checking if the indexed results match the JS codes from doing RPC is easy) 
 (3) hopefully educate others how to do some non-trivial stuff with this auto-generated JS code
We took out Docker deliverables so we're not doing "an additional" product (thank you!) in favor of this more meaningful verifiability testing.  We kept the XCM Transfers / Messages in Milestone 3 because we think its super important to do for the ecosystem.  For focus reasons, we opt to leave out EVM-chain verifiability to a later day, but might attempt it if we're ahead of schedule.  Let me know what you think.

laboon commented on applications/Polkaholic.md:
```diff
@@ -0,0 +1,214 @@
+| 0A. | License | GNUv3 |
+| 0B. | Documentation | We will provide: (a) **inline documentation** of the core crawling and indexing processes (b) how to setup polkaholic.io from an empty Google Cloud project |
+| 1. | UI Augmentation: Browser-based JS for Verifiability | For all key pages, we will add browser-based code to verify results by utilizing RPC calls. |  
+| 2. | UI Augmentation: Server-side JS for Verifiability   | For all key pages, we will publish similar server-side code that also verifies results. |  
```
Can you be more specific about what you mean by "key pages" and which RPC calls are available?  Same with the other deliverables - just looking to be more exact so that we can test it appropriately upon milestone delivery.

Other than that, I'd be very happy to approve this. The interface you already have is very useful.

sourabhniyogi replied:
@laboon To get more specific, we mapped the polkaholic.io site / API into approximately 20 "verification functions" here in [this sheet](https://docs.google.com/spreadsheets/d/1w9xIhyHgB05PgXEVDMyazRYkYvBD3ETaIbWb8X2DKYc/edit#gid=323796891).  All the verification functions will just do the polkadot JS API calls RPC in a way that everyone is familiar with -- eg for verifying block data following a path like  [this](https://github.com/colorfulnotion/polkaholic/blob/main/substrate/test/testGetAuthor.js#L18-L26) and then having "true" or "false" come back if the supplied inputs to the verification function match the data returned by the set of RPC call.

laboon replied:
Thank you, all makes sense to me.

alxs commented:
I'd also be happy to support this. I agree you already have a great product, and together with David's feedback the current scope of the grant seems very reasonable.

I'm also sharing here some feedback from someone else at W3F who had a quick look at your website as to their expectations for a multichain explorer. You might find it useful:

> I want to be able to discover all state changes and logic execution in relation to a given state change or function invocation. If someone makes a transaction (could also be the result of on_initialize/on_finalize/&c), it could send XCMs to other chains, which would then be executed, and that execution could send XCMs to other chains, etc, until it eventually terminates (and it must, unless some chains allow free execution). This will basically look like a tree, but the indexer should discover the tree no matter where someone "enters" it. That is, if I give it the originating transaction, it should find all the downstream effects on other chains. However, I should also be able to look at some state change on a chain which could have been the result of receiving one of those messages, and trace it back "up the tree" (or down the root, whichever way your visual is) to the originating call and then construct the tree.

Lastly, you may want to reach out to the Nova team. They've been collaborating on infrastructure for multi-chain transfers with different teams, and you might find some ground for collaboration.

Noc2 reviewed (changes requested):
Sorry, for the late reply and thanks a lot for the update as well as open sourcing it. That’s great! Just to mention it, in general, we definitely want to support the project. Regarding the milestones, I still have a couple of questions/comments:

- Some formatting seems to be off, e.g.: “| 0e. | Blog Post | We will publish blog posts that (a) explains how parachain and dapp developers can use our XCM Transfers API to support their parachain and dapp development (b) shows how simple XCM transfers are monitored with the Polkaholic.io UI |” 
- Testing Guide	and unit tests for the backend would be nice, see 0c. Of the template https://github.com/w3f/Grants-Program/blob/master/applications/application-template.md
- Could you integrate the table into each milestone?

sourabhniyogi commented:
@alxs @laboon  Thanks for the feedback -- we started doing an XCM Timeline UI (A prototype timeline can be seen [here](https://polkaholic.io/timeline/0x1102d5d85d5047bcaf61793d7dc6944c813691c1ce9ced54d0b0470ffc475743) ) to put a natural home to the verification of XCM Messages from your advisements.   We'll see how to take the next steps to make it a tree and go downstream after we get the basic XCM transfers perfected on every chain.  Will reach out to the Nova team soon --

sourabhniyogi commented:
@Noc2 Apologies for the random formatting -- Ok, I added distinct tables (including a Testing Guide) and detailed the verification functions deliverables in a separate sheet.   I also added limitations on verifiability and limited XCM verification / UI to just DMP/UMP/HRMP.  Let me know if you have any other feedback and will address it in short order.

Noc2 reviewed (approved):
Thanks for the update. I’m happy to approve it and support the project. I will also mark the application as ready for review.

laboon reviewed (approved):
Looks good to me and am looking forward to seeing this project progress.

alxs commented:
@sourabhniyogi thanks for implementing the feedback I shared so unbelievably quickly. Your prototype timeline was received with enthusiasm internally!

alxs commented:
Hey @sourabhniyogi, any updates on the first milestone? Unless you plan to deliver soon, could you please submit an [amendment](https://github.com/w3f/Grants-Program#changes-to-a-grant-after-approval) with an updated timeline?

alxs commented:
@sourabhniyogi makes sense, that does indeed sound like something that would be useful to a lot of teams. What would happen to Polkaholic.io in that case? Will you not continue to develop it further?

Feel free to submit an amendment for changes to the milestone plan in the same way as for the timeline.

Sorry to hear that and I understand your frustration, but my guess as to why your proposal did not pass is as good as yours. You'd have to try to contact council members directly if you really want their feedback. Alternatively, you may want to try again once Gov2 has been rolled out.

sourabhniyogi commented:
@alxs I submitted an amendment via PR #1286 -- we think its a serious improvement over our previous proposal, but please let us know your feedback.   

We will chart a strategy for polkaholic.io on gov1/gov2 after meeting some fellows in Lisbon hopefully, thank you for your support!
